src/main/java/com/shapesecurity/salvation2/Policy.java

```diff
@@ -0,0 +1,1126 @@
+package com.shapesecurity.salvation2;
+
+import com.shapesecurity.salvation2.Directives.FrameAncestorsDirective;
+import com.shapesecurity.salvation2.Directives.HostSourceDirective;
+import com.shapesecurity.salvation2.Directives.PluginTypesDirective;
+import com.shapesecurity.salvation2.Directives.ReportUriDirective;
+import com.shapesecurity.salvation2.Directives.SandboxDirective;
+import com.shapesecurity.salvation2.Directives.SourceExpressionDirective;
+import com.shapesecurity.salvation2.URLs.GUID;
+import com.shapesecurity.salvation2.URLs.URI;
+import com.shapesecurity.salvation2.URLs.URLWithScheme;
+import com.shapesecurity.salvation2.Values.Hash;
+import com.shapesecurity.salvation2.Values.Host;
+import com.shapesecurity.salvation2.Values.MediaType;
+import com.shapesecurity.salvation2.Values.RFC7230Token;
+import com.shapesecurity.salvation2.Values.Scheme;
+
+import javax.annotation.Nonnull;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Policy {
+	// Things we don't preserve:
+	// - Whitespace
+	// - Empty directives or policies (as in `; ;` or `, ,`)
+	// Things we do preserve:
+	// - Source-expression lists being genuinely empty vs consisting of 'none'
+	// - Case (as in lowercase vs uppercase)
+	// - Order
+	// - Duplicate directives
+	// - Unrecognized directives
+	// - Values in directives which forbid them
+	// - Duplicate values
+	// - Unrecognized values
+
+	private List<NamedDirective> directives = new ArrayList<>();
+
+	private SourceExpressionDirective baseUri = null;
+	private boolean blockAllMixedContent = false;
+	private SourceExpressionDirective formAction = null;
+
+	private FrameAncestorsDirective frameAncestors = null;
+
+	private SourceExpressionDirective navigateTo = null;
+
+	private PluginTypesDirective pluginTypes;
+
+	private RFC7230Token reportTo;
```

Comment:
This one's an especially concerning case of the above since it's not even possible to combine `report-to A; report-to B` into a single use of `report-to`.

Reply:
`report-to A; report-to B` is the same as `report-to A`.